airframe-control/src/main/scala/wvlet/airframe/control/CircuitBreaker.scala

```diff
@@ -248,34 +306,46 @@ case class CircuitBreaker(
     * Execute the body block through the CircuitBreaker.
     *
     * If the state is OPEN, this will throw CircuitBreakerOpenException (fail-fast). The state will move to HALF_OPEN state
-    * after a cetain amount of delay, determined by the delayAfterMarkedDead policy.
+    * after a certain amount of delay, determined by the delayAfterMarkedDead policy.
     *
     * If the state is HALF_OPEN, this method allows running the code block once, and if the result is successful,
     * the state will move to CLOSED. If not, the state will be OPEN again.
     *
     * If the state is CLOSED, the code block will be executed normally. If the result is marked failure or nonRetryable exception
-    * is thrown, it will report to the failure to the HealthCheckPolicy. If this policy determins the target service is dead,
+    * is thrown, it will report to the failure to the HealthCheckPolicy. If this policy determines the target service is dead,
     * the circuit will shift to OPEN state to block the future execution.
     *
     */
-  def run[A](body: => A): Unit = {
+  def run[A: ClassTag](body: => A): A = {
     verifyConnection
 
     val result = Try(body)
+
     val resultClass = result match {
-      case Success(x) => resultClassifier(x)
-      case Failure(RetryableFailure(e)) =>
-        ResultClass.retryableFailure(e)
-      case Failure(e) =>
-        errorClassifier(e)
+      case Success(x)                   => resultClassifier(x)
+      case Failure(RetryableFailure(e)) => ResultClass.retryableFailure(e)
+      case Failure(e)                   => errorClassifier(e)
     }
+
     resultClass match {
-      case Succeeded =>
+      case ResultClass.Succeeded =>
         recordSuccess
         result.get
-      case Failed(retryable, cause, extraWait) =>
+      case ResultClass.Failed(retryable, cause, _) =>
         recordFailure(cause)
-        throw cause
+        if (retryable) {
+          // If the error is retryable, rethrow as it is then the caller (maybe Retryer) should handle it.
+          throw cause
+        } else {
+          // If the error is not retryable, apply fallbackHandler
+          val x     = fallbackHandler(cause)
+          val clazz = implicitly[ClassTag[A]].runtimeClass
+          if (clazz.isAssignableFrom(x.getClass)) {
+            x.asInstanceOf[A]
+          } else {
+            throw new ClassCastException(s"${x} is not an instance of ${resultClass}")
```

Comment:
The error message needs to show how to fix it:
`Fallback handler is returning ${x}, which is not an instance of ...`

Reply:
Fixed.